### DISCUSSION TITLE
📝 将 nonebot2 升级到 2.0.0-beta.1 的准备工作

### DISCUSSION THREAD

**Discussion by Lancercmd** [nonebot/discussions#74] [经验]
## 💬 对于 OneBot V11 (CQHTTP)
整理了自有项目所涉及到的变动，作为他人的升级参考
> 涉及的范围仅限 CQHTTP ，欢迎补充更多，不过该讨论的时效性估计也不长
### .env.* 文件
- `loguru.logger` 日志等级与 `debug` 解绑，新增 `log_level` 字段可自定义日志等级，日志等级见 [loguru.logger#levels](https://loguru.readthedocs.io/en/stable/api/logger.html#levels)
```python
# nonebot.config.Config
log_level: Union[int, str] = "INFO"
"""
- **类型**: ``Union[int, str]``
- **默认值**: ``INFO``

:说明:

  配置 NoneBot 日志输出等级，可以为 ``int`` 类型等级或等级名称，参考 `loguru 日志等级`_。

:示例:

.. code-block:: default

    LOG_LEVEL=25
    LOG_LEVEL=INFO

.. _loguru 日志等级:
    https://loguru.readthedocs.io/en/stable/api/logger.html#levels
"""
```
```powershell
# .env.prod
log_level = INFO
```
- `access_token` 字段名变更为 `onebot_access_token`
```powershell
# .env.prod
onebot_access_token = ""
```
### 驱动器
- `Adapter` 的结构有变化，比如获取名称的地方

  > 可能不会用到，但你应该知道
```python
# 升级前
from nonebot.adapters.cqhttp import Bot

Bot.type  # "cqhttp"

# 升级后
from nonebot.adapters.onebot.v11 import Adapter

Adapter.get_name()  # "OneBot V11"
```
- 既然 `Adapter` 名称变了，反向 `websocket`  路由当然也有变化

  > 对 OneBot 实现侧配置的改动
```python
# 升级前
ws://host:port/cqhttp/ws/

# 升级后
ws://host:port/onebot/v11/ws
```
### 注册驱动器
- `Adapter` 的结构有变化，所以注册驱动器的方法也有变化
```python
# 升级前
from nonebot.adapters.cqhttp import Bot as CQHTTPBot

driver.register_adapter('cqhttp', CQHTTPBot)

#升级后
from nonebot.adapters.onebot.v11 import Adapter as OneBot_V11_Adapter

driver.register_adapter(OneBot_V11_Adapter)
```
### 插件文件
- 既然 `Adapter` 名字都改了，那么 import 的位置自然也就变了
```python
# 升级前
from nonebot.adapters.cqhttp import Bot, Event, Message

# 升级后
from nonebot.adapters.onebot.v11 import Bot, Event, Message
```
- ~~由于引入了依赖注入，既有事件函数所传入的参数 `state` 所需的改动~~（nonebot 2.0.0-beta.2 已不需改动此项）
```python
# 升级前
@worker.handle()
async def _(bot: Bot, event: Event, state: T_State) -> None:
    ...


# 升级后
from nonebot.params import State

@worker.handle()
async def _(bot: Bot, event: Event, state: T_State = State()) -> None:
    ...
```
- `event.message` 的变更，对于 `on_command` 事件响应器，现在 `event.message` 会返回含有 `command` 本身的消息

  > 取决于开发习惯，这个改动可能会很费事
  >
  > 在 2.0.0-beta.1 的 `state` 中新增了 `command_arg` 字段，存储了去除 `command` 后的消息
  >
  > 在 2.0.0-beta.1 中也提供了自定义 `event` 的方法，这个到时候看文档去吧
```python
from nonebot.adapters.onebot.v11 import MessageEvent
from nonebot.plugin import on_command

worker = on_command('测试')


@worker.handle()
async def _(event: MessageEvent) -> None:
    print(event.get_plaintext())
```
```python
# 升级前
> 测试abc
abc

# 升级后
> 测试abc
测试abc
```
- `on_command` 事件响应器现在默认 `block=False` 了
- 对于拥有连续会话的功能，现在 `Matcher.got()` 获取并保存到字段的消息将会是 `Message` 类型而不是 `str` 类型

  > 取决于开发习惯，这个改动可能会很费事
  > 
  > 可以通过直接套一个 str() 来回避问题，不过麻烦程度也取决于改动量
```python
@worker.got('key', prompt=MessageTemplate('哈哈'))
async def _(state: T_State = State()) -> None:
    print(state.get('key'))
```
```python
# 升级前
哈哈
> 笑死
笑死

# 升级后
哈哈
> 笑死
...  # 总之是个 Message 对象
```

**Comment by he0119** [comment]
# args_parser

`beta1` 与 `a16`，还有一个区别在于 `args_parser`。在 `beta1` 中，`got` 装饰器 删除了 `args_parser` 参数，同时 `matcher` 也删除了 `args_parser` 装饰器。开发者需要通过 `got` 装饰器中的新参数 `parameterless` 实现类似功能。

## 具体区别

以前的 `got` 会添加两个 `handler`(`_key_getter` 与 `_key_parser` )。`_key_parser` 只会在 `_key_getter` 获取到 `key` 的信息之后运行。如果 `_key_getter` 因为已经获取到 `key` 跳过时，会设置 `_skip_key`，这时 `_key_parser` 会根据 `_skip_key` 同时跳过，也就意味着 `args_parser` 会被跳过，并不会执行。

并且因为是独立的 `handler`，你在 `args_parser` 调用 `matcher.reject()` 只会重新运行 `_key_parser`。

https://github.com/nonebot/nonebot2/blob/cfd992b6b11be61798bf95519e193d578b7efe01/nonebot/matcher.py#L422-L447

但是现在 `got` 会将 `_key_getter` 依赖函数与参数中 `parameterless` 中的依赖函数添加到 `handler` 中。`parameterless` 中的依赖函数会在 `_key_getter` 运行之后运行。但不同的是，`_key_getter` 如果跳过时，你添加到 `parameterless` 中的依赖函数并不会被同时跳过，所以需要额外判断。

因为是通过 `parameterless` 添加到 `handler` 中，所以你在 `parameterless` 依赖函数中调用 `matcher.reject()` 将会重新运行之前所有的 `parameterless` 依赖函数。

https://github.com/nonebot/nonebot2/blob/95415659cee9c8db31f457c42dedf6786ef05923/nonebot/matcher.py#L467-L479

## 区别举例

### a16

```python
matcher = on_command("test")

async def parse_int(bot: Bot, event: Event, state: T_State):
    """处理参数，转换成数字"""
    print("运行 parse_int")
    args = str(event.get_message()).strip()

    # 检查输入参数是不是数字
    if args.isdigit():
        state[state["_current_key"]] = int(args)
    else:
        await matcher.reject("请只输入数字，不然我没法理解呢！")

@matcher.handle()
async def handle_first_receive(bot: Bot, event: Event, state: T_State):
    args = str(event.get_message()).strip()
    if args and args.isdigit():
        state["number"] = int(args)
    
@matcher.got("number", prompt="请输入一个数字。", args_parser=parse_int)
@matcher.got("number2", prompt="请再输入一个数字。", args_parser=parse_int)
async def handle_group_message(bot: Bot, event: Event, state: T_State):
    await matcher.send(str(state["number"] + state["number2"]))
```

可实现如下对话

```plaintxt
群友：/test
机器人: 请输入一个数字。
群友：1
> 运行 parse_int
机器人: 请再输入一个数字。
群友：aaaaa
> 运行 parse_int
机器人: 请只输入数字，不然我没法理解呢！
群友: 2
> 运行 parse_int
机器人: 3
```

或这样，因为在之前的 `handler` 中设置了 `number` 的值，所以 `key` 为 `number` 的 `got` 被自动跳过，`parse_int` 函数也就不会执行。

```plaintxt
群友：/test 3
机器人: 请再输入一个数字。
群友：aaaaa
> 运行 parse_int
机器人: 请只输入数字，不然我没法理解呢！
群友: 2
> 运行 parse_int
机器人: 5
```

### b1

```python
matcher = on_command("test")

def parse_int(key: str):
    """解析数字，并将结果存入 state 中"""

    async def _key_parser(
        matcher: Matcher, state: T_State = State(), input: Union[int, Message] = Arg(key)
    ):
        print(f"运行 parse_int, key: {key}")
        if isinstance(input, int):
            return

        plaintext = input.extract_plain_text()
        if not plaintext.isdigit():
            await matcher.reject_arg(key, "请只输入数字，不然我没法理解呢！")
        state[key] = int(plaintext)

    return _key_parser

@matcher.handle()
async def handle_first_receive(state: T_State = State(), args: Message = CommandArg()):
    plaintext = args.extract_plain_text()
    if plaintext.isdigit():
        state["number"] = int(plaintext)
    
@matcher.got("number", prompt="请输入一个数字。", parameterless=[Depends(parse_int("number"))])
@matcher.got("number2", prompt="请再输入一个数字。", parameterless=[Depends(parse_int("number2"))])
async def handle_group_message(number: int = Arg(), number2: int = Arg()):
    await matcher.send(str(number + number2))
```

无论是否提前设置了 `number` 的值，均会执行 `parse_int` 这个依赖函数。

```plaintxt
群友：/test
机器人: 请输入一个数字。
群友：1
> 运行 parse_int, key: number
机器人: 请再输入一个数字。
群友：aaaaa
> 运行 parse_int, key: number
> 运行 parse_int, key: number2
机器人: 请只输入数字，不然我没法理解呢！
群友: 2
> 运行 parse_int, key: number
> 运行 parse_int, key: number2
机器人: 3
```

```plaintxt
群友：/test 3
> 运行 parse_int, key: number
机器人: 请再输入一个数字。
群友：aaaaa
> 运行 parse_int, key: number
> 运行 parse_int, key: number2
机器人: 请只输入数字，不然我没法理解呢！
群友: 2
> 运行 parse_int, key: number
> 运行 parse_int, key: number2
机器人: 5
```

## 适配示例

以我自己的禁言插件为例（这只是其中一种修改方法，比较适合我自己之前写的代码）

将

https://github.com/he0119/CoolQBot/blob/eb77046dd9f8c53c4e7b2e8419d2e447261ade97/src/plugins/ban/__init__.py#L71-L80

https://github.com/he0119/CoolQBot/blob/eb77046dd9f8c53c4e7b2e8419d2e447261ade97/src/plugins/ban/__init__.py#L101

修改为

https://github.com/he0119/CoolQBot/blob/4c15529663f22efd885884d6c74a93794c012508/src/utils/helpers.py#L61-L75

https://github.com/he0119/CoolQBot/blob/4c15529663f22efd885884d6c74a93794c012508/src/plugins/ban/__init__.py#L83-L87

**Comment by ffreemt** [comment]
感谢花时间写贴，节省了不少时间。

补充一条。

基于poetry的插件升级到nonebot2 beta：
```bash
poetry update  # 或 poetry install nonebot2@latest ？
poetry add nonebot-adapter-onebot
poetry remove nonebot-adapter-cqhttp
```
或基于pip
```bash
pip install nonebot2 -U
pip install nonebot-adapter-onebot
pip uninstall nonebot-adapter-cqhttp
```
